### PR TITLE
feat!: Update audit_config submodule to support multiple log types

### PR DESCRIPTION
**Pull request by agnnn**
Update modules/audit_config to support multiple log types for a specified service while defining audit logs config.
Make use of dynamic block on resource deployment.

Fixes:
1.  Supports multiple log types for a service.
2.  Added README file with example usage.

### CODE REVIEW

**Comment by morgante** [comment]
@agnnn is it necessary to have each service only listed once? I'm wondering if the existing implementation would work with a config like this:

```
  audit_log_config = [
    {
      service    = "pubsub.googleapis.com"      
      log_config = {
          log_type = DATA_READ
          exempted_members = [
            "group:my-group@my-org.com",
            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
            "user:my-user@my-org.com"
          ]
        },
  },
  {
      service    = "pubsub.googleapis.com" 
      log_config = {
          log_type = DATA_WRITE
          exempted_members = [
            "group:my-group@my-org.com",
            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
            "user:my-user@my-org.com"
          ]
    }
  }]
```

**Comment by agnnn** [comment]
@morgante So basically... each resource is authoritative and not additive. When we have log_type for a single service defined on multiple resources, it will apply only one configuration. Every time doing a terraform apply.. it goes into a loop applying changes for that particular service that got modified due to the last terraform apply.
This is the same case for the existing implementation and the new implementation.
Because of this same limitation, the proposed implementation now supports having multiple log types defined for a single service on a single resource.
So yes, we need to specify each service only once.

**Comment by morgante** [comment]
Got it, thanks.

I *think* we could actually accomplish this without changing the module's variable interface, it will just require some more logic in the module. The logic would be something like:

1. Gather a list of all unique services (ex. using distinct()) that were provided in the variable.
2. For_each on each service
3. On each service, add the audit configs listed for that service.

**Comment by kpeder** [comment]
> I'd like to take the above approach for 2 reasons:
> 
> 1. Allowing this change to be backwards-compatible
> 2. Providing a simpler (less nested) interface to module users.

re: 1. - It is a breaking change, but on the other hand, the current implementation is broken.
re: 2. - Con: interface input is not DRY

**Comment by morgante** [comment]
> re: 1. - It is a breaking change, but on the other hand, the current implementation is broken.

The current implementation isn't broken. It works fine for users who only want 1 consistent config—this PR is an enhancement, not a bug fix.

**Comment by kpeder** [comment]
> > re: 1. - It is a breaking change, but on the other hand, the current implementation is broken.
> 
> The current implementation isn't broken. It works fine for users who only want 1 consistent config—this PR is an enhancement, not a bug fix.

Arguably, and I am not arguing particularly hard here, the initial implementation assumes an additive resource vs an authoritative one. Since the underlying resource supports multiple log types per service, and since the current implementation can't configure more than one log type per service, I consider it broken for a (not so arguably, since each service supports 3 different audit configs) very common use case.

Functionally, the important part is addition of the dynamic block, and this is why I'm not going to argue too hard. I will note that the current implementation is released and can be pinned to preserve the simpler use case.

**Comment by agnnn** [comment]
@morgante Any updates on this ?? Or do you still wanna go with the approach you suggested earlier ??

**Comment by morgante** [comment]
@agnnn Yes, I'd like this to be updated to the approach I suggested (which will be backwards-compatible).

**Comment by kpeder** [comment]
So, I have done some investigation on how this could be implemented, and the conversion works with this pattern:

`variable "audit_log_config" {
  type = list(object({ service : string, log_type : string, exempted_members : list(string) }))
  default = [
    {
      service          = "pubsub.googleapis.com"      
      log_type         = "DATA_READ"
      exempted_members = [
            "group:my-group@my-org.com",
            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
            "user:my-user@my-org.com"
          ]
    },
    {
      service          = "pubsub.googleapis.com" 
      log_type         = "DATA_WRITE"
      exempted_members = [
            "group:my-group@my-org.com",
            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
            "user:my-user@my-org.com"
          ]
    }
  ]
}

locals {
  audit_log_config = {
    for key, val in var.audit_log_config :
    val.service => val...
  }
}

output "test" {
  value = local.audit_log_config
}`

**Comment by thiagonache** [comment]
@kpeder my understanding is that you can go ahead with your approach. Since you keep the variable and the same and it still works with just one item, we are good. Feel free to tag me before asking for Morgante's final review.
I'd encourage you to run tests before pushing it to make sure it will pass.

**Comment by agnnn** [comment]
@morgante Hey, so I have updated the PR based on your requested changes and it is almost ready to be reviewed except for the failing check, could you inspect the logs and suggest what is the error on lint ??

**Comment by thiagonache** [comment]
hi @agnnn let me see if I can reproduce the error.

**Review by bharathkkb — COMMENTED** [review]
Thanks for the PR @agnnn 
Please run `make generate_docs` to regenerate docs. You can also run `make docker_test_lint` to test link check locally.
Errors from CI
```
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=.git' '--exclude=autogen' '--exclude=*.tfvars' /workspace/modules/audit_config/README.md /tmp/tmp.K5k2UROctV/generate_docs/workspace/modules/audit_config/README.md
49,51c49,50
< | audit_log_config | List of objects to be added to audit log config | list(object({ service : string, log_type : string, exempted_members : list(string) })) | n/a | yes |
< | project | GCP Project ID | string | n/a | yes |
< 
---
> | audit\_log\_config | List of objects to be added to audit log config | object | n/a | yes |
> | project | Project to add the IAM policies/bindings | string | n/a | yes |
57,58c56
< | audit_log_config | Map of log type and exempted members added to service |
< 
---
> | audit\_log\_config | Map of log type and exempted members to be added to service |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
Checking for trailing whitespace
./modules/audit_config/README.md:11:  
./modules/audit_config/README.md:14:      service          = "pubsub.googleapis.com"      
./modules/audit_config/README.md:23:      service          = "storage.googleapis.com" 
./modules/audit_config/README.md:32:      service          = "pubsub.googleapis.com" 
Error: Trailing whitespace found in the lines above.

Checking for missing newline at end of file
Running shellcheck
Checking file headers
Running flake8
Running terraform fmt
Running terraform validate
terraform_validate ./examples/billing_account 
Success! The configuration is valid.
```

**Comment by thiagonache** [comment]
@agnnn seems to be an issue in the CI, not related to your PR. Everything passes locally on my machine. Meanwhile, I'll try to help and give it my review.

**Comment by thiagonache** [comment]
@bharathkkb that's not hapenning 

> @agnnn seems to be an issue in the CI, not related to your PR. Everything passes locally on my machine. Meanwhile, I'll try to help and give it my review.

🤦  I ran tests from the master branch. My bad.

**Comment by thiagonache** [comment]
@bharathkkb It looks good to me. I deployed with the latest release and then deployed this branch and everything still working perfectly.

(Sorry for previous mistakes. I'll do better on paying attention to the details)

**Comment by agnnn** [comment]
@bharathkkb Thanks for your quick reply. So I did run `make docker_generate_docs` as well as `make docker_test_lint` and pushed the commit.

```
make docker_generate_docs
docker run --rm -it \
        -v /mnt/c/Users/Ravindran/Desktop/work/terraform-google-iam:/workspace \
        gcr.io/cloud-foundation-cicd/cft/developer-tools:0.12.0 \
        /bin/bash -c 'source /usr/local/bin/task_helper_functions.sh && generate_docs'
Generating markdown docs with terraform-docs
Working in ./examples/billing_account ...
Success!
Working in ./examples/custom_role_org ...
Success!
Working in ./examples/custom_role_project ...
Success!
Working in ./examples/folder ...
Success!
Working in ./examples/kms_crypto_key ...
Success!
Working in ./examples/kms_key_ring ...
Success!
Working in ./examples/member_iam ...
Success!
Working in ./examples/organization ...
Success!
Working in ./examples/project ...
Success!
Working in ./examples/project_conditions ...
Success!
Working in ./examples/pubsub_subscription ...
Success!
Working in ./examples/pubsub_topic ...
Success!
Working in ./examples/service_account ...
Success!
Working in ./examples/stackdriver_agent_roles ...
Success!
Working in ./examples/storage_bucket ...
Success!
Working in ./examples/subnet ...
Success!
**Working in ./modules/audit_config ...
Success!**
Working in ./modules/billing_accounts_iam ...
Success!
Working in ./modules/custom_role_iam ...
Success!
Working in ./modules/folders_iam ...
Success!
Skipping ./modules/helper because README.md does not exist.
Working in ./modules/kms_crypto_keys_iam ...
Success!
Working in ./modules/kms_key_rings_iam ...
Success!
Working in ./modules/member_iam ...
Success!
Working in ./modules/organizations_iam ...
Success!
Working in ./modules/projects_iam ...
Success!
Working in ./modules/pubsub_subscriptions_iam ...
Success!
Working in ./modules/pubsub_topics_iam ...
Success!
Working in ./modules/service_accounts_iam ...
Success!
Working in ./modules/storage_buckets_iam ...
Success!
Working in ./modules/subnets_iam ...
Success!
Skipping ./test/fixtures/additive because README.md does not exist.
Skipping ./test/fixtures/authoritative because README.md does not exist.
Skipping ./test/fixtures/billing-iam because README.md does not exist.
Skipping ./test/fixtures/custom-role because README.md does not exist.
Skipping ./test/fixtures/helper because README.md does not exist.
Skipping ./test/fixtures/helper/base because README.md does not exist.
Skipping ./test/fixtures/member-iam because README.md does not exist.
Skipping ./test/fixtures/static-and-dynamic because README.md does not exist.
Skipping ./test/fixtures/static-and-dynamic/static_projects because README.md does not exist.
Skipping ./test/setup because README.md does not exist.
```

```
make docker_test_lint
docker run --rm -it \
        -v /mnt/c/Users/Ravindran/Desktop/work/terraform-google-iam:/workspace \
        gcr.io/cloud-foundation-cicd/cft/developer-tools:0.12.0 \
        /usr/local/bin/test_lint.sh
Checking for documentation generation
Checking for trailing whitespace
Checking for missing newline at end of file
Running shellcheck
Checking file headers
Running flake8
Running terraform fmt
Running terraform validate
terraform_validate ./examples/billing_account
Success! The configuration is valid.

terraform_validate ./examples/custom_role_org
Success! The configuration is valid.

terraform_validate ./examples/custom_role_project
Success! The configuration is valid.

terraform_validate ./examples/folder
Success! The configuration is valid.

terraform_validate ./examples/kms_crypto_key
Success! The configuration is valid.

terraform_validate ./examples/kms_key_ring
Success! The configuration is valid.

terraform_validate ./examples/member_iam
Success! The configuration is valid.

terraform_validate ./examples/organization
Success! The configuration is valid.

terraform_validate ./examples/project
Success! The configuration is valid.

terraform_validate ./examples/project_conditions
Success! The configuration is valid.

terraform_validate ./examples/pubsub_subscription
Success! The configuration is valid.

terraform_validate ./examples/pubsub_topic
Success! The configuration is valid.

terraform_validate ./examples/service_account
Success! The configuration is valid.

terraform_validate ./examples/stackdriver_agent_roles
Success! The configuration is valid.

terraform_validate ./examples/storage_bucket
Success! The configuration is valid.

terraform_validate ./examples/subnet
Success! The configuration is valid.

**terraform_validate ./modules/audit_config
Success! The configuration is valid.**

terraform_validate ./modules/billing_accounts_iam
Success! The configuration is valid.

terraform_validate ./modules/custom_role_iam
Success! The configuration is valid.

terraform_validate ./modules/folders_iam
Success! The configuration is valid.

terraform_validate ./modules/helper
Success! The configuration is valid.

terraform_validate ./modules/kms_crypto_keys_iam
Success! The configuration is valid.

terraform_validate ./modules/kms_key_rings_iam
Success! The configuration is valid.

terraform_validate ./modules/member_iam
Success! The configuration is valid.

terraform_validate ./modules/organizations_iam
Success! The configuration is valid.

terraform_validate ./modules/projects_iam
Success! The configuration is valid.

terraform_validate ./modules/pubsub_subscriptions_iam
Success! The configuration is valid.

terraform_validate ./modules/pubsub_topics_iam
Success! The configuration is valid.

terraform_validate ./modules/service_accounts_iam
Success! The configuration is valid.

terraform_validate ./modules/storage_buckets_iam
Success! The configuration is valid.

terraform_validate ./modules/subnets_iam
Success! The configuration is valid.

terraform_validate ./test/fixtures/additive
Success! The configuration is valid.

terraform_validate ./test/fixtures/authoritative
Success! The configuration is valid.

terraform_validate ./test/fixtures/billing-iam
Success! The configuration is valid.

terraform_validate ./test/fixtures/custom-role
Success! The configuration is valid.

terraform_validate ./test/fixtures/helper
Success! The configuration is valid.

terraform_validate ./test/fixtures/helper/base
Success! The configuration is valid.

terraform_validate ./test/fixtures/member-iam
Success! The configuration is valid.

terraform_validate ./test/fixtures/static-and-dynamic
Success! The configuration is valid.

terraform_validate ./test/fixtures/static-and-dynamic/static_projects
Success! The configuration is valid.

terraform_validate ./test/setup
Success! The configuration is valid.
```
Everything looks good on my local system. But on the PR is still failing the check. Any suggestions ??

File: modules/audit_config/README.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+## Example Usage
+```
+module "audit_log_config" {
+  source           = "terraform-google-modules/iam/google//modules/audit_config"
+
+  project          = my-project
+  
+audit_log_config = [
+    {
+      service          = "pubsub.googleapis.com"
```
Comment (bharathkkb) [inline modules/audit_config/README.md:14]:
```suggestion
      service          = "pubsub.googleapis.com"
```

File: modules/audit_config/README.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      service          = "pubsub.googleapis.com"      
+      log_type         = "DATA_READ"
+      exempted_members = [
+            "group:my-group@my-org.com",
+            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
+            "user:my-user@my-org.com"
+          ]
+    },
+    {
+      service          = "storage.googleapis.com"
```
Comment (bharathkkb) [inline modules/audit_config/README.md:23]:
```suggestion
      service          = "storage.googleapis.com"
```

File: modules/audit_config/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      service          = "storage.googleapis.com" 
+      log_type         = "DATA_WRITE"
+      exempted_members = [
+            "group:my-group@my-org.com",
+            "serviceAccount:my-sa@my-project.iam.gserviceaccount.com",
+            "user:my-user@my-org.com"
+          ]
+    },
+    {
+      service          = "pubsub.googleapis.com"
```
Comment (bharathkkb) [inline modules/audit_config/README.md:32]:
```suggestion
      service          = "pubsub.googleapis.com"
```

File: modules/audit_config/README.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+This optional module is used to configure audit log configs for a project.
+
+## Example Usage
+```
+module "audit_log_config" {
+  source           = "terraform-google-modules/iam/google//modules/audit_config"
+
+  project          = my-project
+
```
Comment (bharathkkb) [inline modules/audit_config/README.md:11]:
```suggestion

```

**Review by bharathkkb — COMMENTED** [review]
Not sure why your local CI is passing but `Checking for trailing whitespace`, I have highlighted the lint nits below

**Comment by agnnn** [comment]
@bharathkkb Done, All checks are passing now. This Looks good to go for a review.

**Review by bharathkkb — APPROVED** [review]
Overall LGTM
Since the keys are changing there will be a delete recreate which I think is okay for a minor release.

**Comment by thiagonache** [comment]
@bharathkkb the key is not changing. Just the logic.

**Comment by bharathkkb** [comment]
@thiagonache IIUC the change
```diff
- key => val
+ val.service => val...
```
will change the key from an array index like `0` to a string like `pubsub.googleapis.com` which cause TF to want to delete previous resource instances and recreate them. Now thinking about it, it might even need two TF apply's as the first one may error out due to TF wanting to the create and delete concurrently as each is authoritative.

**Comment by thiagonache** [comment]
> @thiagonache IIUC the change
> 
> ```diff
> - key => val
> + val.service => val...
> ```
> 
> will change the key from an array index like `0` to a string like `pubsub.googleapis.com` which cause TF to want to delete previous resource instances and recreate them. Now thinking about it, it might even need two TF apply's as the first one may error out due to TF wanting to the create and delete concurrently and each is authoritative.

My understanding is since the value is variadic, the state won't change if you don't change the input. And I actually tested it out, if the input is the same there's no changes on the terraform plan.

**Comment by bharathkkb** [comment]
@thiagonache I just tested it and it produced a diff with the new keys; for instance destroy `google_project_iam_audit_config.project["0"]` and  create `google_project_iam_audit_config.project["storage.googleapis.com"]`. Curious why you did not observe this.

full diff: https://gist.github.com/bharathkkb/3f48d70a015c5c76ebc751259b601493
config:
```hcl
{
    service          = "storage.googleapis.com"
    log_type         = "DATA_READ"
    exempted_members = ["serviceAccount:ci-iam-member-0-5da1@project-id.iam.gserviceaccount.com"]
},
{
    service          = "allServices"
    log_type         = "DATA_READ"
    exempted_members = ["serviceAccount:ci-iam-member-0-5da1@project-id.iam.gserviceaccount.com"]

  }
```

**Comment by thiagonache** [comment]
> @thiagonache I just tested it and it produced a diff with the new keys; for instance destroy `google_project_iam_audit_config.project["0"]` and create `google_project_iam_audit_config.project["storage.googleapis.com"]`. Curious why you did not observe this.
> 
> full diff: https://gist.github.com/bharathkkb/3f48d70a015c5c76ebc751259b601493
> config:
> 
> ```terraform
> {
>     service          = "storage.googleapis.com"
>     log_type         = "DATA_READ"
>     exempted_members = ["serviceAccount:ci-iam-member-0-5da1@project-id.iam.gserviceaccount.com"]
> },
> {
>     service          = "allServices"
>     log_type         = "DATA_READ"
>     exempted_members = ["serviceAccount:ci-iam-member-0-5da1@project-id.iam.gserviceaccount.com"]
> 
>   }
> ```

@bharathkkb Idk what I've done and how you still don't send me to hell and stop to bother you. I'm bad at reviews :( I wanna do it more to get better, I hope you don't mind.
As to apologies I've created the patch to be applied in @agnnn branch to fix this and works as discussed in the issue. Here is the [link](https://gist.github.com/thiagonache/02f994e20921022d168fcf28a11ae318)

**Comment by bharathkkb** [comment]
@thiagonache np :) Regarding your modifications, I am still not sure why we need to keep the array indices as keys. Doing so will create unnecessary diffs if for instance I modify the order of the input list of objects?

**Comment by thiagonache** [comment]
@bharathkkb  Because I use the index to group configs but keep the previous order by getting distinct apis. It was suggested by Morgante

**Comment by bharathkkb** [comment]
@thiagonache ack, IIUC having the `local.api_services` as `index => service` serves mostly as a way to prevent that delete recreate. IMHO we should just delete recreate and transition to using services as keys as it would still keep the interface same and added benefit of not being dependent on order of objects in `audit_log_config`.

**Comment by thiagonache** [comment]
That was the initial idea but @morgante didn’t like it.
I would suggest to change it on the next major change and add the feature now with this workaround.

**Comment by kpeder** [comment]
The real point if the update is around ensuring more than one configuration can be applied per service api.

I would suggest to update test inputs and assertions to configure multiple log type configuration is happening properly.

IMO the destroy/recreate isn't a big issue... even on destroy its only a configuration update under the hood; no resource is being removed.

**Comment by thiagonache** [comment]
Well, it may be a big deal to be without audit in production even for a short time. If we need to force recreate it should be a breaking change

**Comment by thiagonache** [comment]
As a module user, I wouldn't be happy with my resources being recreated in a minor release.

**Comment by thiagonache** [comment]
> Got it, thanks.
> 
> 
> 
> I *think* we could actually accomplish this without changing the module's variable interface, it will just require some more logic in the module. The logic would be something like:
> 
> 
> 
> 1. Gather a list of all unique services (ex. using distinct()) that were provided in the variable.
> 
> 2. For_each on each service
> 
> 3. On each service, add the audit configs listed for that service.
> 
> 
> 
> I'd like to take the above approach for 2 reasons:
> 
> 1. Allowing this change to be backwards-compatible
> 
> 2. Providing a simpler (less nested) interface to module users.

@kpeder @bharathkkb this is why I coded my patch.

**Comment by agnnn** [comment]
Hey @bharathkkb from the above comments from @kpeder  and @thiagonache  what is your suggestion, if you're planning to do a major release, I assume it should be okay to recreate resources.

**Comment by kpeder** [comment]
> Hey @bharathkkb from the above comments from @kpeder and @thiagonache what is your suggestion, if you're planning to do a major release, I assume it should be okay to recreate resources.

I would endorse Thiago's patch; and open an issue to track and add the breaking change as is here on next major release?

**Comment by bharathkkb** [comment]
Historically we have done minor releases for some safe delete recreates like IAM.  Ack that even a short time without audit is not ideal in prod so we can do a major release too.

**Comment by morgante** [comment]
I'm fine with making this a breaking change if we think the delete/recreate is a major issue (even though it's only for a tiny amount of time).

Can we add some test cases though? I'm really not sure if this implementation works the way I was looking for.

**Pull request merged**